experiment/telegram/telegram.go

```diff
@@ -18,27 +22,101 @@ const (
 )
 
 // Config contains the experiment config.
-type Config struct {
-	// LogLevel is the MK log level. Empty implies "WARNING".
-	LogLevel string
+type Config struct{}
+
+// TestKeys contains telegram test keys.
+type TestKeys struct {
+	TelegramHTTPBlocking bool                           `json:"telegram_http_blocking"`
+	TelegramTCPBlocking  bool                           `json:"telegram_tcp_blocking"`
+	TelegramWebFailure   *string                        `json:"telegram_web_failure"`
+	TelegramWebStatus    *string                        `json:"telegram_web_status"`
+	TCPConnect           []ootemplate.TCPConnectResults `json:"tcp_connect"`
+	Requests             []ootemplate.HTTPRoundTrip     `json:"requests"`
 }
 
 func measure(
 	ctx context.Context, sess *session.Session, measurement *model.Measurement,
 	callbacks handler.Callbacks, config Config,
 ) error {
-	settings := measurementkit.NewSettings(
-		"Telegram", sess.SoftwareName, sess.SoftwareVersion,
-		sess.CABundlePath(), sess.ProbeASNString(), sess.ProbeCC(),
-		sess.ProbeIP(), sess.ProbeNetworkName(), config.LogLevel,
-	)
-	out, err := measurementkit.StartEx(settings, sess.Logger)
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+	testkeys := &TestKeys{
+		TelegramHTTPBlocking: true,
+		TelegramTCPBlocking:  true,
+	}
+	measurement.TestKeys = testkeys
+	defer func() {
+		for _, tcpconn := range testkeys.TCPConnect {
+			if tcpconn.Status.Success == true {
+				testkeys.TelegramTCPBlocking = false
+				break
+			}
+		}
+		for _, roundtrip := range testkeys.Requests {
+			if strings.Contains(roundtrip.Request.URL, "web.telegram.org") {
+				continue
+			}
+			if roundtrip.Failure == "" {
+				testkeys.TelegramHTTPBlocking = false
+				break
+			}
+		}
+		var failure *string
+		count := 0
+		for _, roundtrip := range testkeys.Requests {
```

Comment:
You also need to check if the HTML title contains the string "Telegram Web", see: https://github.com/measurement-kit/measurement-kit/blob/master/src/libmeasurement_kit/ooni/telegram.cpp#L101.

Reply:
Thank you, you're right. I ported this change to #76.